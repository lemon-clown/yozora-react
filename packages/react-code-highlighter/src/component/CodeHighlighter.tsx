import PropTypes from 'prop-types'
import React from 'react'
import vscDarkTheme from '../theme/vsc-dark'
import vscLightTheme from '../theme/vsc-light'
import type { IPrismTheme } from '../types/prism'
import { HighlightContent } from './HighlightContent'

export interface ICodeHighlighterProps {
  /**
   * Source code contents
   */
  value: string
  /**
   * Code language
   */
  lang?: string | null
  /**
   * Line number of Lines that should be highlighted.
   */
  highlightLinenos?: number[]
  /**
   * Whether the code block is in a collapsed state.
   * @default false
   */
  collapsed?: boolean
  /**
   * Maximum number of rows displayed
   * @default -1
   */
  maxLines?: number
  /**
   * Line height.
   * @default '1.6rem'
   */
  lineHeight?: React.CSSProperties['lineHeight']
  /**
   * Whether should display line numbers.
   * @default true
   */
  showLineNo?: boolean
  /**
   * If true, use vscDarkTheme as default theme,
   * otherwise use vscLightTheme as default theme.
   * @default true
   */
  darken?: boolean
  /**
   * Highlight prism theme.
   */
  theme?: IPrismTheme
  /**
   * Ref of the codes area.
   */
  codesRef?: React.RefCallback<HTMLDivElement> | React.RefObject<HTMLDivElement>
  /**
   * Callback when linenoWidth changed.
   */
  onLinenoWidthChange?(linenoWidth: React.CSSProperties['width']): void
}

export const CodeHighlighter: React.FC<ICodeHighlighterProps> = props => {
  const {
    lang,
    value: code,
    darken = true,
    highlightLinenos,
    maxLines = -1,
    lineHeight = '1.6rem',
    collapsed = false,
    showLineNo = true,
    codesRef,
    onLinenoWidthChange,
  } = props

  const theme: IPrismTheme = props.theme ?? (darken ? vscDarkTheme : vscLightTheme)
  return (
    <HighlightContent
      code={code}
      codesRef={codesRef}
      collapsed={collapsed}
      highlightLinenos={highlightLinenos}
      lang={lang ?? undefined}
      lineHeight={lineHeight}
      maxLines={maxLines}
      showLineNo={showLineNo}
      theme={theme}
      onLinenoWidthChange={onLinenoWidthChange}
    />
  )
}

CodeHighlighter.propTypes = {
  codesRef: PropTypes.any,
  collapsed: PropTypes.bool,
  darken: PropTypes.bool,
  highlightLinenos: PropTypes.arrayOf(PropTypes.number) as React.Validator<
    number[] | null | undefined
  >,
  lang: PropTypes.string,
  lineHeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  maxLines: PropTypes.number,
  onLinenoWidthChange: PropTypes.func,
  showLineNo: PropTypes.bool,
  theme: PropTypes.any,
  value: PropTypes.string.isRequired,
}
CodeHighlighter.displayName = 'YozoraCodeHighlighter'
