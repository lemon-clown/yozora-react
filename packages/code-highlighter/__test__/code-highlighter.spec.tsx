import { mount, render } from 'enzyme'
import React, { useEffect, useState } from 'react'
import CodeHighlighter from '../src'

describe('basic rendering case', () => {
  const errorLogger = jest.spyOn(global.console, 'error').mockImplementation((...args) => {
    throw new Error(args.join(' '))
  })

  afterAll(() => {
    errorLogger.mockRestore()
  })

  it('test lineno change', () => {
    function Wrapper(): React.ReactElement {
      const [code, setCode] = useState<string>('let a: number = 1 + 2;')

      useEffect(() => {
        const nextCode =
          'let a = 1, b = 2\n' +
          Array.from(new Array(100))
            .map((x, i) => '// ' + i)
            .join('\n') +
          "\nlet c = 3\nconsole.log('c:', c)"
        setCode(nextCode)
      }, [])

      return (
        <pre data-line-count={code.split(/\r\n|\n|\r/g).length}>
          <CodeHighlighter lang="typescript" value={code} />
        </pre>
      )
    }

    const wrapper = mount(<Wrapper />)
    expect(wrapper.getDOMNode().getAttribute('data-line-count')).toEqual(String(103))

    const lines = wrapper.find({ linenoWidth: '2.5em' })
    for (let i = 0; i < lines.length; ++i) {
      const line = lines[i]
      const s = getComputedStyle(line.getDOMNode())
      expect(s.width).toEqual('3.5em')
    }
  })

  it('snapshot', () => {
    const wrapper = render(
      <pre>
        <CodeHighlighter lang="typescript" value="let a: number = 1 + 2;" />
      </pre>,
    )
    expect(wrapper).toMatchSnapshot()
  })
})
