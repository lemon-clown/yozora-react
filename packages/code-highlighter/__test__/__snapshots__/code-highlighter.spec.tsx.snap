// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic rendering case snapshot 1`] = `
<pre>
  <div
    class="yozora-code-highlighter prism-code language-typescript"
    style="color:#d4d4d4;background-color:none;line-height:1.6rem;max-height:3.2rem;min-height:100%"
  >
    <div
      class="yozora-code-highlighter__linenos"
      style="width:2.2em"
    >
      <div
        class="yozora-code-highlighter__line"
      >
        <span>
          1
        </span>
      </div>
    </div>
    <div
      class="yozora-code-highlighter__codes"
    >
      <div
        class="yozora-code-highlighter__codes-wrapper"
      >
        <div
          class="yozora-code-highlighter__line token-line"
          style="color:#d4d4d4"
        >
          <span
            class="token keyword"
            style="color:rgb(86, 156, 214)"
          >
            let
          </span>
          <span
            class="token plain"
          >
             a
          </span>
          <span
            class="token operator"
            style="color:rgb(212, 212, 212)"
          >
            :
          </span>
          <span
            class="token plain"
          >
             
          </span>
          <span
            class="token builtin"
            style="color:rgb(79, 193, 255)"
          >
            number
          </span>
          <span
            class="token plain"
          >
             
          </span>
          <span
            class="token operator"
            style="color:rgb(212, 212, 212)"
          >
            =
          </span>
          <span
            class="token plain"
          >
             
          </span>
          <span
            class="token number"
            style="color:rgb(181, 206, 168)"
          >
            1
          </span>
          <span
            class="token plain"
          >
             
          </span>
          <span
            class="token operator"
            style="color:rgb(212, 212, 212)"
          >
            +
          </span>
          <span
            class="token plain"
          >
             
          </span>
          <span
            class="token number"
            style="color:rgb(181, 206, 168)"
          >
            2
          </span>
          <span
            class="token punctuation"
            style="color:rgb(106, 153, 85)"
          >
            ;
          </span>
        </div>
      </div>
    </div>
  </div>
</pre>
`;
